### DISCUSSION TITLE
Writing generic functions for Zod Objects

### DISCUSSION THREAD

**Discussion by DouglasdeMoura** [colinhacks/zod#2595] [Q&A]
I'm trying to write a generic function that receives a Zod Schema, call the `.omit({ id: true })` on it and return it. Similarly to the example on the [documentation](https://github.com/colinhacks/zod#writing-generic-functions), I've tried this:

```typescript
import { z } from 'zod'

function makeCreateSchema<T extends z.AnyZodObject>(schema: T) {
  return schema.omit({ id: true })
}

const schema = z.object({
  id: z.string(),
  name: z.string()
})

const createSchema = makeCreateSchema(schema)

type CreateSchema = z.infer<typeof createSchema>
```

Unfortunately, this does not give me the correct type for the `CreateSchema`:

![image](https://github.com/colinhacks/zod/assets/564542/95e2fe12-859e-42b2-a0b5-d1b5d46083f5)

So, is it possible to do what I'm trying to achieve?

**Comment by kylecarhart** [comment]
Bumping this. I'm running into this same thing right now and having trouble finding an example of functions that take in a zod object as a parameter and perform an `omit` on it.
